### PR TITLE
github: Add pull request template.

### PR DESCRIPTION
**Pull request by neiljp**
**What does this PR do?**

Adds a pull request template, like I'm using here :)

Take a look at the raw and rendered versions to get a full idea of the contributor experience.

- [X] New Feature?
- [X] Fixes Bug? Partially #830
- [ ] Internal Refactor?

**Tested?** 
- [X] Manually 
- [ ] Tests adapted (if necessary)
- [ ] Tests added/changed for any new behavior
- [ ] Passed linting & tests (overall PR)
- [ ] Passed linting & tests (each commit)

**Notes/Doubts/Concerns/Questions** 

This is more detailed than others I've seen, eg. https://raw.githubusercontent.com/zulip/zulip/master/.github/pull_request_template.md

Do we need the top title? Or feature/bug boxes? (except for suggesting what it may fix?)

Do the testing boxes help? Do they just take up space?

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<!-- Please see https://github.com/zulip/zulip-terminal#contributor-guidelines ! -->
+
+**What does this PR do?**
+
+Overall description goes here!
+
+- [ ] New Feature?
+- [ ] Fixes Bug? [Fixes <issue>; Partially <issue> | Fixes point <n> of <issue> | Resolves an issue that's not filed]
```
Comment (zee-bit) [inline .github/pull_request_template.md:8]: Perhaps also include the discussion link, if there's no issue filed for the fix?

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+- [ ] Internal Refactor?
+
+**Tested?** <!-- Fine to leave some of these unchecked! -->
+- [ ] Manually
+- [ ] Tests adapted (if necessary)
+- [ ] Tests added/changed for any new behavior
```
Comment (zee-bit) [inline .github/pull_request_template.md:14]: Does the "adapted" and "changed" carry the same meaning here? Then we could remove the first check-box?

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+- [ ] Passed linting & tests (overall PR)
+- [ ] Passed linting & tests (each commit)
+
+<!-- See https://github.com/zulip/zulip-terminal#commit-style -->
+**Commit flow** <!-- if more than one commit; add/delete/fill-in as appropriate -->
+* first commit doing some thing
```
Comment (zee-bit) [inline .github/pull_request_template.md:20]: Although I've used this field in some of my own PR's, I wonder if it's really necessary?
The "Commit flow" is essentially just the commit headers, as we have documented in that link, which can be easily accessed from the commits tab at the top?

File: .github/pull_request_template.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+* maybe multiple commits doing similar things
+* ...
+
+**Notes/Doubts/Concerns/Questions** <!-- if any; add/delete/fill-in as appropriate -->
```
Comment (zee-bit) [inline .github/pull_request_template.md:24]: This field is great. :+1: 
Can we simplify the title to just "Notes/Concerns"?
Doubts/Concerns/Questions sounds pretty similar to me.

**Review by zee-bit — COMMENTED** [review]
Thanks for getting this done, Neil. It already looks amazing. :+1: 

This is going to be very helpful, particularly, for new contributors. It will also be helpful for reviewers in getting to know what the PR does (abstractly), and how ready it is - is it properly tested, does the linting pass are the test added yet. Reviewers can also provide suggestions for any "Notes/Concerns" that the author has mentioned in the PR template without looking at the code - if possible too.

File: .github/pull_request_template.md (hunk 1 above)
Comment (neiljp) [reply .github/pull_request_template.md:8]: Good point, though I think that's a separate entry to suggest including.

File: .github/pull_request_template.md (hunk 2 above)
Comment (neiljp) [reply .github/pull_request_template.md:14]: The intent is to ensure that existing tests still pass, via slight changes if necessary - which is distinct from covering different new behavior. Perhaps the two should be explicitly "Existing tests pass (adapted if necessary)" and "New tests added (if new behavior)". Maybe they're too long though.

File: .github/pull_request_template.md (hunk 4 above)
Comment (neiljp) [reply .github/pull_request_template.md:24]: I agree 4 is too many. Maybe just "Notes" and push the possible content into the examples.

Of course the titles are editable/removable, but people may not think that, so simple titles may be easier as users are more likely to think to at least fill in their own content I hope :)

File: .github/pull_request_template.md (hunk 3 above)
Comment (neiljp) [reply .github/pull_request_template.md:20]: I agree the commit list can be seen in the PR itself - and does change over time, often, and it's not necessary to keep updating the text here as that happens.

I do feel we could benefit from having something here though - maybe others would like to weigh in.

File: .github/pull_request_template.md (hunk 2 above)
Comment (Rohitth007) [reply .github/pull_request_template.md:14]: Maybe we could follow the same wording as in the commit style guide?
Tests added (new feature)
Tests updated (refactor/feature that needs minor amending in existing tests)

**Comment by neiljp** [comment]
I'm more inclined to remove the first 3 checkboxes as GitHub includes them as a list of tasks - which they are not. I'm not sure there's an easy alternative, but we can simply hint at these elements in the description prompt, or in the commit flow perhaps too?

File: .github/pull_request_template.md (hunk 2 above)
Comment (Ezio-Sarthak) [reply .github/pull_request_template.md:14]: @Rohitth007 I think that'd require the contributor to explicitly write details in parenthesis? I believe we're keener on the tests in the PR in general, and specific additions/amendments are addressed in commits in any case.

Comment (Rohitth007) [reply .github/pull_request_template.md:14]: @Ezio-Sarthak No, I just used the parenthesis for explaining added vs updated here

File: .github/pull_request_template.md (hunk 3 above)
Comment (neiljp) [reply .github/pull_request_template.md:20]: @Rohitth007 That's what the examples are intended to indicate, ie. that it's a summary. I meant to include that they're only ~ commit titles, but I don't want to motivate just copy and pasting those either.

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+**Notes & Questions** <!-- if any; add/delete/fill-in as appropriate -->
+<!-- For example:
+- this doesn't include feature X (yet?)
+- unsure about Y
+- should this do Z?
+-->
+
+**Interactions** <!-- if any; add/delete/fill-in as appropriate -->
+- Waiting on other PR(s)? What numbers?
+- Blocks other PR(s)? What numbers?
```
Comment (preetmishra) [inline .github/pull_request_template.md:32]: Should these two points exist as a comment like the above two sections?

**Review by preetmishra — COMMENTED** [review]
@neiljp Thanks for working on this! :+1: This looks great. I have left a minor in-line comment.

**Pull request merged**